man/pcapmerge.1
.\" yo there.
.TH PCAPMERGE 1  "22 Dec 2002"
.SH NAME
pcapmerge \- merge multiple pcap files into one larger one
.SH SYNOPSIS
.na
.B pcapmerge
[
.B -V
] [
.B -h
]
.B \-o
.I outfile
pcap1 pcap2 ...
.br
.ad
.SH DESCRIPTION
.LP
.I Pcapmerge
is a program for creating a single pcap (tcpdump) file from two or more pcap files.  The
resulting file contains the packets from each of the source pcap files in the order they
are specified on the command line.
.SH OPTIONS
.LP
.TP
.B \-h
Help
.TP
.B \-o <file>
Output filename.  The name of the file to write.
.TP
.B \-V
Print version info and exit.
.SH "SEE ALSO"
tcpdump(8), tcpreplay(8), capinfo(1)
.SH AUTHOR
Matt Bing
.LP
The current version is packaged with tcpreplay which is available via HTTP:
.LP
.RS
.I http://www.sourceforge.net/projects/tcpreplay/
.RE


Docs/CHANGELOG
$Id: CHANGELOG 1390 2005-07-03 19:59:48Z aturner $

07/03/2005: Version 2.3.5
    - Fix crash with replay live mode (-S)
    - Fix longstanding tcpprep auto/router mode bug

06/05/2005: Version 2.3.4
    - Support libpcap <= 0.5 for OpenBSD
    - libpcap 0.5 doesn't have pcap_open_dead, so disable -w, -W and -D in
      that case
    - Fix configure errors regarding libpcapnav
    - Fix compile problem in libpcap.c with OpenBSD
    - Fix compile problem where libpcap and libnet are in /usr and
      libpcapnav is somehwere else
    - Don't ship .svn directories
    - Close stdin so we can reopen it if necessary

02/09/2005: Version 2.3.3
    - Fix port rewriting feature on little-endian systems
    - configure now properly handles --with-libnet and --with-libpcap

11/08/2004: Version 2.3.2
    - When sending via -1, report which interface the packet will exit
    - Fix bug when caplen > packet len
    - Allow rewriting of Layer 2 via -2 for Cisco HDLC (DLT_CHDLC)

09/19/2004: Version 2.3.1
    - Fix bug with fakepcap.c which appeared on systems using an older
      version of libpcap (such as Red Hat 9.0)
    - Don't die when setting STDERR to non-blocking

09/05/2004: Version 2.3.0
    - Fix longstanding endian bug in cache files on little endian systems
      (note that this breaks compatibility w/ existing cache files created
      on little endian systems)
    - Add support to tcpreplay and tcpprep for DLT_CHDLC (Cisco HDLC)
    - Clean up validate_l2() and rewrite_l2()
    - Write a simple perl script to parse net/bpf.h of DLT values
    - Teach everything the names of all the current DLT values
    - Detect if libpcap supports pcap_datalink_val_to_description()
    - Start printing datalink descriptions instead of DLT values
    - Remove magic numbers from tcpreplay.c
    - Add a HACKING document

06/21/2004: Version 2.2.2
    - tcpprep now supports DLT_RAW and DLT_LINUX_SLL
    - add makefile target for website docs (FAQ.html, FAQ.pdf, CHANGELOG)
    - Fix some sanity checks in tcpreplay for processing various DLT types
      in validate_l2()
    - Fix -x & -X
    - Merge in patch from Denis which rewrites TCP/UDP ports via -4
    - Fix rewrite of source MAC address in single interface mode (bug #975848)

05/16/2004: Version 2.2.1
    - Fix compile issue under RH9
    - Fix compile issue when not using --with-debug

05/15/2004: Version 2.2.0
    - Fix pseudo-NAT (not evaluating all rules and an infinate loop)
    - Start using strtok_r() in any function to prevent future bugs
    - Minor updates to tcpprep.1 & tcpreplay.8 man pages
    - Re-org some functions into different files for better modularity
    - Clean up of some of the cache comment code
    - flowreplay man page moved to section 1
    - Update tcpprep and tcpreplay man pages and the FAQ
    - Improve documentation regarding pseudo-NAT feature
    - Fix one output mode which treated all packets as primary
    - Add endpoint mode (-e) which rewrites all traffic between two IP's
    - Fix rewrite of IP addresses in ARP requests & replies w/ pseudo-NAT
    - Fix CIDR matching of 0.0.0.0/0 (all packets) which matched only 
      255.255.255.255
    - All CIDR notation now accepts IP addresses w/o requiring /32
    - non-debug mode now uses -O3 -funroll-loops for better performance

05/01/2004: Version 2.1.1
    - Fix ntohll/htonll compile error on big endian systems

04/23/2004: Version 2.1.0
    - Add support for per output interface/file NAT tables 
    - Add support for using dual output features w/ a single output
    - Add support to tcpprep for splitting via destination port
    - Now fully 64bit when tracking number of packets
    - Fix a bug where sometimes the last few packets are not sent when using
      a tcpprep cache file
    - Some code refactorization/cleanup
    - tcpprep cache files now support user comments
    - Fix bug where regex optimization was turned always turned off

03/24/2004: Version 2.0.3
    - Add support for rewriting src mac & Linux SLL loopback frames
    - Update FAQ

02/25/2004: Version 2.0.2
    - Fix compile issue in edit_packet.c on strict aligned archs

02/03/2004: Version 2.0.1
    - Re-organize FAQ and add more content
    - Add support for "pseudo NAT" (-N) for ARP and IPv4
    - Code optimization to only run the checksum fixer once per packet
    - Clean up help (-h) a little

02/01/2004: Version 2.0.0
    - Remove libpcapnav requirement
    - Now support libpcapnav >= 0.4
    - Add -1 to replay one packet at a time (user must hit <ENTER>)
    - Add tcpdump packet parsing to print packets as sent (-v)
    - Place flowreplay manpage in correct location
    - More FAQ updates
    - Rename 1.5.x as 2.0
    - Fix/standardize all licensing info.  Still BSD of course.
    - -T now forces -F
    - tcpprep now actually accepts -n (client|server)
    - Update the INSTALL doc
    - Remove the Docs/README... the FAQ has replaced it.

12/10/2003: Version 1.5.alpha6
    - Add BPF filter support to tcpprep and tcpreplay (-x F:"filter")
    - Update the FAQ
    - Add two new auto modes to tcpprep (client and server)
    - Make clean no longer wipes out the compiled documentation in Docs
    - Add support for replaying live traffic
    - Add bridge mode
    - Add -L to limit the total number of packets to send

11/03/2003: Version 1.5.alpha5
    - Add -T to truncate packets > MTU so they can be sent
    - Now fixes ICMP checksums as appropriate
    - Updated FAQ
    - Updated flowreplay design doc
    - Merge packetrate code from 1.4.5
    - Fix compile issues under Libnet 1.1.1
    - --with-debug now enables debuging during 'make test'
    - Fix various Solaris compatibility bugs
    - Add data dump mode which dumps layer 7 data to the file (-D)
    - Now requires libpcapnav
    - Allow to jump X bytes into the pcap and start replaying packets (-o)
    - Can now split traffic/data into files (-w & -W)

07/16/2003: Version 1.5.alpha4

pcapmerge.c
/* $Id: pcapmerge.c 767 2004-10-06 12:48:49Z aturner $ */

/*
 * Copyright (c) 2001-2004 Aaron Turner, Matt Bing.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the names of the copyright owners nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* The pcapmerge available at http://indev.insu.com/Fwctl/pcapmerge.html
 * is written in Perl and does not handle large pcaps very well. This
 * is meant to be small and efficient.
 * 
 * gcc -o pcapmerge pcapmerge.c -lpcap
 * 
 * usage: pcapmerge -o bigpcap pcap1 pcap2 pcap3 .. 
 *
 * Tested on OpenBSD 3.1. Probably won't work on non BSD.
 */

#include <pcap.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/time.h>
#include <sys/types.h>

#include "config.h"
#include "tcpreplay.h"
#include "err.h"
#include "queue.h"

/* we get this from libpcap */
extern char pcap_version[];

#ifdef DEBUG
int debug = 0;
#endif

/* magic constants for various pcap file types */
#define PCAP_MAGIC          		0xa1b2c3d4
#define PCAP_SWAPPED_MAGIC      	0xd4c3b2a1
#define PCAP_MODIFIED_MAGIC     	0xa1b2cd34
#define PCAP_SWAPPED_MODIFIED_MAGIC     0x34cdb2a1

#define SWAPLONG(y) \
((((y)&0xff)<<24) | (((y)&0xff00)<<8) | (((y)&0xff0000)>>8) | (((y)>>24)&0xff))

#define SWAPSHORT(y) \
( (((y)&0xff)<<8) | ((u_short)((y)&0xff00)>>8) )

/* data prefixing each packet in modified pcap */
struct pcap_mod_pkthdr {
    struct pcap_pkthdr hdr;     /* normal header */
    u_int32_t ifindex;          /* receiving interface index */
    u_int16_t protocol;         /* ethernet packet type */
    u_int8_t pkt_type;          /* ethernet packet type */
    u_int8_t pad;               /* padding */
};

/* info about each open file */
struct pcap_file {
    int fd;
    int modified;
    int swapped;
    char *name;
      SLIST_ENTRY(pcap_file) next;
};

void init_files(int, char **);
void write_packets(struct pcap_file *, int);
void usage();

SLIST_HEAD(, pcap_file) files;
     struct pcap_file_header hdr;
     int outfd;
     char *outfile;


     void
       version()
{
    fprintf(stderr, "pcapmerge version: %s\n", VERSION);
    fprintf(stderr, "Compiled against libpcap: %s\n", pcap_version);
    exit(0);
}

int
main(int argc, char *argv[])
{
    struct pcap_file *p;
    int ch, fd;

    outfile = NULL;

    while ((ch = getopt(argc, argv, "o:h?V")) != -1) {
        switch (ch) {
        case 'o':              /* output file */
            outfile = optarg;
            break;
        case 'V':
            version();
            break;
        default:
            usage();
        }
    }

    argc -= optind;
    argv += optind;

    if (outfile == NULL)
        errx(1, "must specify output file");

    SLIST_INIT(&files);
    init_files(argc, argv);

    if ((fd = open(outfile, O_WRONLY | O_CREAT, 0644)) < 0)
        err(1, "open %s", outfile);

    /* write file header */
    if (write(fd, &hdr, sizeof(hdr)) != sizeof(hdr))
        err(1, "write");

    SLIST_FOREACH(p, &files, next) {
        write_packets(p, fd);
    }

    return 0;
}

/* Read all the file headers and set the pcap_file_header for output file */
void
init_files(int len, char *filelist[])
{
    struct pcap_file_header phdr;
    struct pcap_file *file;
    int i, fd, modified, swapped;
    u_int32_t snaplen, linktype;

    snaplen = linktype = 0;

    for (i = 0; i < len; i++) {
        if ((fd = open(filelist[i], O_RDONLY, 0)) < 0) {
            warn("skipping %s: could not open", filelist[i]);
            close(fd);
            continue;
        }

        if (read(fd, (void *)&phdr, sizeof(phdr)) != sizeof(phdr)) {
            warn("skipping %s: could not read header", filelist[i]);
            close(fd);
            continue;
        }

        switch (phdr.magic) {
        case PCAP_MAGIC:
            swapped = 0;
            modified = 0;
            break;

        case PCAP_SWAPPED_MAGIC:
            swapped = 1;
            modified = 0;
            break;

        case PCAP_MODIFIED_MAGIC:
            swapped = 0;
            modified = 1;
            break;

        case PCAP_SWAPPED_MODIFIED_MAGIC:
            swapped = 1;
            modified = 1;
            break;

        default:
            warn("skipping %s: invalid pcap", filelist[i]);
            close(fd);
            continue;
        }

        /* ensure everything is in host-byte order */
        if (swapped) {
            phdr.snaplen = SWAPLONG(phdr.snaplen);
            phdr.linktype = SWAPLONG(phdr.linktype);
        }

        if (linktype == 0) {
            linktype = phdr.linktype;
        }
        else if (linktype != phdr.linktype) {
            warn("skipping %s: inconsistent linktype", filelist[i]);
            close(fd);
            continue;
        }

        /* pick the largest snaplen */
        if (phdr.snaplen > snaplen)
            snaplen = phdr.snaplen;

        if ((file = (struct pcap_file *)malloc(sizeof(*file))) == NULL)
            errx(1, "out of memory");

        file->fd = fd;
        file->modified = modified;
        file->swapped = swapped;
        file->name = filelist[i];

        SLIST_INSERT_HEAD(&files, file, next);
    }

    /* set the header for the output file */
    hdr.magic = PCAP_MAGIC;
    hdr.version_major = 2;
    hdr.version_minor = 4;
    hdr.thiszone = 0;
    hdr.sigfigs = 0;
    hdr.snaplen = snaplen;
    hdr.linktype = linktype;
}

void
write_packets(struct pcap_file *p, int out)
{
    struct pcap_pkthdr p1, *phdr;
    struct pcap_mod_pkthdr p2;
    char pkt[MAXPACKET];
    int len, ret;

    for (;;) {
        if (p->modified) {
            ret = read(p->fd, &p2, sizeof(p2));
            if (ret == -1)
                err(1, "read");
            else if (ret == 0)
                break;
            phdr = &p2.hdr;
        }
        else {
            ret = read(p->fd, &p1, sizeof(p1));
            if (ret == -1)
                err(1, "read");
            else if (ret == 0)
                break;
            phdr = &p1;
        }

        if (p->swapped) {
            len = SWAPLONG(phdr->caplen);
            phdr->ts.tv_sec = SWAPLONG(phdr->ts.tv_sec);
            phdr->ts.tv_usec = SWAPLONG(phdr->ts.tv_usec);
            phdr->caplen = SWAPLONG(phdr->caplen);
            phdr->len = SWAPLONG(phdr->len);
        }
        else {
            len = phdr->caplen;
        }

        if (len > MAXPACKET) {
            warn("skipping %s: abnormally large packet (%d bytes)", p->name,
                 len);
            break;
        }

        ret = read(p->fd, pkt, len);
        if (ret == -1)
            err(1, "read");
        else if (ret == 0) {
            warn("skipping %s: truncated packet3", p->name);
            break;
        }

        if (write(out, phdr, sizeof(*phdr)) != sizeof(*phdr))
            err(1, "write");
        if (write(out, pkt, len) != len)
            err(1, "write");

    }

    close(p->fd);
}

void
usage()
{
    fprintf(stderr, "usage: pcapmerge -o bigpcap pcap1 pcap2 pcap3 ..\n");
    exit(1);
}
